GameProject/Assets/TurnBasedAssets/Scripts/Dialogue/RenderDialogue.cs

```diff
@@ -0,0 +1,127 @@
+﻿using System;
+using System.Collections;
+using System.Collections.Generic;
+using System.Net.Mime;
+using System.Text;
+using UnityEngine;
+using UnityEngine.UI;
+
+namespace TurnBasedAssets.Scripts.Dialogue
+{
+    public class RenderDialogue : MonoBehaviour
+    {
+        [SerializeField] private Text _pageName;
+        [SerializeField] private Text _pageText;
+        [SerializeField] private GameObject _pageImagePosition;
+        [SerializeField] private float _sentenceSpeed;
+        [SerializeField] private Button _option;
+        [SerializeField] private GameObject[] _buttonPositions;
+        [SerializeField] private Image _dialogueBackground;
+        [SerializeField] private GameObject _dialogueBox;
+        private List<Button> _responseOptions = new List<Button>();
+        private Image _previousImage;
+        private NpcMoods _npcImageMoods;
+        private Image _newMoodImage;
+        private Coroutine _currentRoutine;
+        private bool IsPreviousImageNotNull => _previousImage != null;
+        private bool IsCurrentRoutineNotNull => _currentRoutine != null;
+        private bool IsNewMoodImageNotNull => _newMoodImage != null;
+
+        private void RenderPageText(string pageName, string pageText)
+        {
+            _pageName.text = pageName;
+            _pageText.text = pageText;
+        }
+
+        private IEnumerator Play(Dialogue npc, Message npcMessage)
+        {
+            var sb = new StringBuilder();
+            var letters = npcMessage.MessageText.ToCharArray();
+            foreach (var letter in letters)
+            {
+                sb.Append(letter);
+                RenderPageText(npc.NpcName, sb.ToString());
+                yield return new WaitForSeconds(_sentenceSpeed);
+            }
+            yield return null;
+        }
+
+        public void PlayParagraphCycle(Dialogue npcDialogue, NpcImages npcImages, int paragraphNumber)
+        {
+            if (paragraphNumber < 0)
```

Comment:
have you considered just `return`ing after calling `EndDialogue()` here? It'd reduce your scope count by one.

Reply:
I don't understand?